src/main/java/com/google/devtools/build/lib/skyframe/SequencedSkyframeExecutor.java

```diff
@@ -513,14 +519,20 @@ private void handleDiffsWithMissingDiffInformation(
         diffPackageRootsUnderWhichToCheck.add(pair.getFirst());
       }
 
-      EnumSet<FileType> fileTypesToCheck =
-          EnumSet.of(FileType.EXTERNAL_REPO, FileType.EXTERNAL_IN_MANAGED_DIRECTORY);
+      EnumSet<FileType> fileTypesToCheck = EnumSet.noneOf(FileType.class);
+      if (checkExternalFiles) {
+        fileTypesToCheck =
+            EnumSet.of(FileType.EXTERNAL_REPO, FileType.EXTERNAL_IN_MANAGED_DIRECTORY);
+        if (externalFilesKnowledge.tooManyNonOutputExternalFilesSeen) {
+          fileTypesToCheck.add(FileType.EXTERNAL);
+        }
+      }
       // See the comment for FileType.OUTPUT for why we need to consider output files here.
       if (checkOutputFiles) {
         fileTypesToCheck.add(FileType.OUTPUT);
       }
-      if (externalFilesKnowledge.tooManyNonOutputExternalFilesSeen) {
-        fileTypesToCheck.add(FileType.EXTERNAL);
+      if (!checkExternalFiles && !checkOutputFiles && diffPackageRootsUnderWhichToCheck.isEmpty()) {
+        throw new AssertionError("should have already early exited");
```

Comment:
nit: It's a little more idiomatic IMO to do a Preconditions.checkState call here